### PR TITLE
CRAFTERCMS-3070

### PR DESCRIPTION
**Pull request by dejan-brkic**
Added Repository API v2

### Ticket reference or full description of what's in the PR
https://github.com/craftercms/craftercms/issues/3070

### CODE REVIEW

File: src/main/api/2/studio-api-2.yaml [hunk 1]
```diff
@@ -1456,6 +1460,235 @@ paths:
...
+              properties:
+                siteId:
+                  type: string
+                  description: Site ID
+                remoteName:
+                  type: string
+                  description: Remote repository name to pull from
+                remoteBranch:
+                  type: string
+                  description: Remote repository branch to pull from
```
Comment (sumerjabri) [inline src/main/api/2/studio-api-2.yaml:1787]: Add ability to specify merge strategy: https://www.git-scm.com/docs/git-pull#Documentation/git-pull.txt--Xltoptiongt

File: src/main/api/2/studio-api-2.yaml [hunk 2]
```diff
@@ -1456,6 +1460,235 @@ paths:
...
+                properties:
+                  response:
+                    $ref: '#/components/schemas/ApiResponse'
+        '400':
+          $ref: '#/components/responses/BadRequest'
+        '401':
+          $ref: '#/components/responses/Unauthorized'
+        '404':
+          $ref: '#/components/responses/NotFound'
+        '500':
```
Comment (sumerjabri) [inline src/main/api/2/studio-api-2.yaml:1810]: Add `409` and return the git conflict message. Indicate that the user must contact an admin.

File: src/main/api/2/studio-api-2.yaml [hunk 3]
```diff
@@ -1456,6 +1460,235 @@ paths:
...
```
Comment (sumerjabri) [inline src/main/api/2/studio-api-2.yaml:1858]: Add `409` and indicate git conflict and the user should notify the administrator.

File: src/main/api/2/studio-api-2.yaml [hunk 4]
```diff
@@ -1456,6 +1460,235 @@ paths:
...
+              properties:
+                siteId:
+                  type: string
+                  description: Site ID
+                remoteName:
+                  type: string
+                  description: Remote repository name to push to
+                remoteBranch:
+                  type: string
+                  description: Remote repository branch to push to
```
Comment (sumerjabri) [inline src/main/api/2/studio-api-2.yaml:1836]: Add support for `-f` `--force` to force push to remote. https://www.git-scm.com/docs/git-push#Documentation/git-push.txt--f

File: src/main/api/2/studio-api-2.yaml [hunk 5]
```diff
@@ -1456,6 +1460,235 @@ paths:
...
+        '400':
+          $ref: '#/components/responses/BadRequest'
+        '401':
+          $ref: '#/components/responses/Unauthorized'
+        '404':
+          $ref: '#/components/responses/NotFound'
+        '500':
+          $ref: '#/components/responses/InternalServerError'
+
+  /repository/rebuild_databese:
```
Comment (sumerjabri) [inline src/main/api/2/studio-api-2.yaml]: Please fix typo in `databases`

**Review by sumerjabri — CHANGES_REQUESTED** [review]
Add permissions to all API calls.

File: src/main/api/2/studio-api-2.yaml [hunk 6]
```diff
@@ -2067,6 +2300,65 @@ components:
...
+        remotePassword:
+          type: string
+          description: password to use to access remote repository
+        remoteToken:
+          type: string
+          description: token to use to access remote repository
+        remotePrivateKey:
+          type: string
+          description: private key to access
+    RemoteInfo:
```
Comment (sumerjabri) [inline src/main/api/2/studio-api-2.yaml]: Change to `RemoteRepositoryInfo`

File: src/main/api/2/studio-api-2.yaml [hunk 7]
```diff
@@ -1456,6 +1460,235 @@ paths:
...
+                    $ref: '#/components/schemas/ApiResponse'
+        '400':
+          $ref: '#/components/responses/BadRequest'
+        '401':
+          $ref: '#/components/responses/Unauthorized'
+        '404':
+          $ref: '#/components/responses/NotFound'
+        '500':
+          $ref: '#/components/responses/InternalServerError'
+  /repository/list_remotes:
```
Comment (sumerjabri) [inline src/main/api/2/studio-api-2.yaml:1938]: Add an empty line above.

**Review by sumerjabri — CHANGES_REQUESTED** [review]

**Comment by dejan-brkic** [comment]
Repository API v2

### Ticket reference or full description of what's in the PR
https://github.com/craftercms/craftercms/issues/3070

File: src/main/java/org/craftercms/studio/impl/v2/service/repository/RepositoryManagementServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+    @Override
+    @HasPermission(type = DefaultPermission.class, action = "rebuild_database")
+    public void rebuildDatabase(@ProtectedResourceId(SITE_ID_RESOURCE_ID) String siteId) {
+        siteService.rebuildDatabase(siteId);
+    }
+
+    @Override
+    @HasPermission(type = DefaultPermission.class, action = "remove_remote")
+    public boolean removeRemote(String siteId, String remoteName) throws CryptoException {
```
Comment (avasquez614) [inline src/main/java/org/craftercms/studio/impl/v2/service/repository/RepositoryManagementServiceImpl.java]: Isn't this missing the `@ProtectedResourceId(SITE_ID_RESOURCE_ID)`?

**Review by avasquez614 — CHANGES_REQUESTED** [review]

**Review by sumerjabri — APPROVED** [review]

**Review by avasquez614 — APPROVED** [review]

**Pull request merged**